Add Guess.js article from io19 (#1052)
GoogleChrome/web.dev: src/site/content/en/blog/predictive-prefetching/index.md

]
// ...
```

If you're not using Google Analytics, you can specify a `reportProvider` and download data from your favorite service.

### Integration with frameworks

To learn more about how to integrate Guess.js with your favorite framework check out these resources:

1. [Using Guess.js with Angular](https://guess-js.github.io/docs/angular)
1. [Using Guess.js with Next.js](https://guess-js.github.io/docs/next)
1. [Using Guess.js with Nuxt.js](https://guess-js.github.io/docs/nuxt)

For a quick walkthrough on the integration with Angular, check out this video:

{% YouTube '5FRxQiGqqmM' %}

### How does Guess.js work?

Here's how Guess.js implements predictive prefetching:

1. It first extracts data for the user navigational patterns from your favorite analytics provider.
1. It then maps the URLs from the report to the JavaScript chunks produced by webpack.
1. Based on the extracted data, it creates a simple predictive model of which pages a user is likely to navigate to from any given page.
1. It invokes the model for each JavaScript chunk, predicting the other chunks that are likely to be needed next.
1. It adds prefetching instructions to each chunk.

> In the end, each chunk will contain prefetching instructions similar to:

Reviewer: ```suggestion
When Guess.js is done, each chunk will contain prefetching instructions similar to:
```

```js
__GUESS__.p(
  ['a.js', 0.2],
  ['b.js', 0.8]
)
```

> The meaning behind this code that Guess.js generated is: consider prefetching chunk `a.js` with probability `0.2` and chunk `b.js` with probability `0.8`.

Reviewer: ```suggestion
This Guess.js-generated code is telling the browser to consider prefetching chunk `a.js` with probability `0.2` and chunk `b.js` with probability `0.8`.
```

> Once the browser executes this code Guess.js will first check the user's connection speed. If it's sufficient enough Guess.js will insert two `<link rel="prefetch">` tags in the header of the page, one for each chunk. If the user has a poor network connection, Guess.js will only prefetch the chunk `b.js` since it has `0.8` probability to be needed next.

Reviewer: ```suggestion
Once the browser executes the code, Guess.js will check the user's connection speed. If it's sufficient, Guess.js will insert two `<link rel="prefetch">` tags in the header of the page, one for each chunk. If the user is on a high-speed network, Guess.js will prefetch both chunks. If the user has a poor network connection, Guess.js will only prefetch chunk `b.js` since it has a high probability of being needed.
```
Fleshed this out to try to make the logic a bit more explicit. See what you think.

> ![How Guess.js works](guess-2.png)

Reviewer: This image has several details not covered in the post. Wondering if maybe we should cut it to avoid possible confusion? (For example, probably don't want people thinking, "What the heck is a Markov Chain?")

Author: Do you think there are other details missing than "Markov Chain" and "TensorFlow"?

I can update the image.

---
title: Faster Web Navigations with Predictive Prefetching
subhead: Learn what is predictive prefetching and how Guess.js implements it.
date: 2019-06-26
hero: predictive-prefetching-header.png
alt: Guess.js logo
authors:
  - mgechev
description: |
  Code-splitting allows you to speed up your applications, but it may slow down subsequent-navigations. Predictive prefetching is an efficient way to empower data analytics to smartly prefetch what the user is likely to use next, optimizing the network utilization.
tags:
  - post
  - performance
---

During my session ["Faster Web Navigations with Predictive Prefetching"](https://www.youtube.com/watch?v=0jB4YWgAxUo) at Google I/O 2019, I talked about optimizing web apps with code-splitting and its potential performance implication on subsequent page navigations. In the second part of the talk, I discussed how to improve navigation speed by empowering predictive prefetching and how Guess.js implements this technique.

> <div style="width:100%; padding-top: 56.25%; position: relative;">

Author: I didn't know about the shortcode. I was looking at another blog post as a template and didn't find it. Thanks for pointing me out to it!

## Prefetching JavaScript

![Prefetching example](guess-1.png)

Prefetching is a technique which allows you to download in the background and save into the cache resources that you think the user will need soon. The usual way to do this is by using `<link rel="prefetch">`. Two common pitfalls with prefetching are:

- Prefetching too many resources and consuming a lot of data (*overfetching*)
- Never prefetching the resources the user needs

Predictive allows you to solve these problems by consuming a report for the users' navigational patterns.

## Predictive Prefetching with Guess.js

A JavaScript library which provides predictive prefetching functionality is [Guess.js](https://github.com/guess-js). Guess.js consumes a report from Google Analytics or another analytics provider to build a predictive model that could be used for smartly prefetching only what the user may need in the future.

> Guess.js has integration with Angular, Next.js, Nuxt.js, and Gatsby. To use it in your application, you should add a single line of webpack configuration specifying a [Google Analytics view ID](https://stackoverflow.com/questions/36898103/what-is-a-viewid-in-google-analytics):

Reviewer: Thanks for adding the links! Any concerns with making the edits to the language? They'd improve the alignment with web.dev guidelines for instructions (e.g., by using the imperative mood).

```js/0,5
const { GuessPlugin } = require('guess-webpack');

// ...
plugins: [
   // ...
   new GuessPlugin({ GA: 'XXXXXX' })
]
// ...
```

If you're not using Google Analytics, you can specify a `reportProvider` and download data from your favorite service.

> ### Integration with Angular

Author: Yep, good point. I picked Angular because I spent the most working on the integration.

I'll link to the codelabs from https://guess-js.github.io and keep the video.

> ### Integration with Angular

Reviewer: Awesome! Can you just make it an unordered list? (We save ordered lists for instructions.)

To integrate Guess.js with an Angular CLI project, you can use [`ngx-build-plus`](https://github.com/manfredsteyer/ngx-build-plus). The steps you need to go through include:

* Install of `ngx-build-plus`, `guess-webpack`, and `guess-parser`
* Add the `GuessPlugin` as part of your webpack configuration specifying a Google Analytics view ID or, if using different analytics provider, a `reportProvider`:

```js
const { GuessPlugin } = require('guess-webpack');
const { parseRoutes } = require('guess-parser');

module.exports = {
  plugins: [
    new GuessPlugin({
      GA: 'VIEW_ID',
      runtime: {
	      delegate: false
      },
      routeProvider() {
	      return parseRoutes('.');
      }
    })
  ]
};
```

* Build your application specifying your extended webpack config with the `--extra-webpack-config` option

> For a video introduction, check the video below.

Reviewer: ```suggestion
For a quick walkthrough, check out this video:
```

<div style="width:100%; padding-top: 56.25%; position: relative;">
<iframe style="width:100%; height: 100%;position: absolute; top: 50%; left:
50%; transform: translate(-50%,-50%);"
src="https://www.youtube.com/embed/5FRxQiGqqmM" frameborder="0"
allow="accelerometer; autoplay; encrypted-media; gyroscope; picture-in-picture"
allowfullscreen></iframe>
</div>

## How Does Guess.js Work

> Guess.js will add predictive prefetching going through the following steps:

Reviewer: ```suggestion
Here's how Guess.js implements predictive prefetching:
```
Suggest integrating this chunk into the "Predictive prefetching with Guess.js" section, right under the first paragraph. That way, the sequence goes:
1. What is Guess?
1. How does it work?
1. How can you use it? (via various integrations)

> 1. Extract data from your favorite analytics provider

Reviewer: ```suggestion
1. It first extracts data from your favorite analytics provider.
```
Suggest being more precise about what data is extracted. Like "common navigation patterns" or something?

> 1. Map the URLs from the report to the JavaScript chunks produced by webpack

Reviewer: ```suggestion
1. It then maps the URLs from the report to the JavaScript chunks produced by webpack.
```

> 1. Create a simple predictive model

Reviewer: ```suggestion
1. Based on the extracted data, it creates a simple predictive model of which pages a user is likely to navigate to from any given page.
```
Should this go second?

> 1. Invoke the model for each chunk, predicting which are the other chunks likely to be needed next

Reviewer: ```suggestion
1. It invokes the model for each JavaScript chunk, predicting the other chunks that are likely to be needed next.
```

> 1. Add prefetching instructions in each chunk

Reviewer: ```suggestion
1. It adds prefetching instructions to each chunk.
```
A little more explanation or an example might help folks understand the mechanism here?

Author: :+1: I'll add an image.

Author: @mfriesenhahn I added a paragraph with an explanation to not bloat with another image. Would you verify if it's easier to get?

![How Guess.js works](guess-2.png)

## Learn more

> You learn more about Guess.js in the following resources:

Reviewer: ```suggestion
To learn more about Guess.js, check out these resources:
```

> - [Faster Web Navigations with Predictive Prefetching](https://www.youtube.com/watch?v=0jB4YWgAxUo)

Reviewer: Navigation generally isn't a countable noun (i.e., you don't usually add an "s" to the end). Unless this is a technical usage of the word, is it possible to fix this in the YouTube video title?
